rust/main/agents/relayer/src/server/message_retry.rs

```diff
@@ -1,32 +1,95 @@
 use crate::settings::matching_list::MatchingList;
+
 use axum::{extract::State, routing, Json, Router};
+
 use derive_new::new;
-use tokio::sync::broadcast::Sender;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{broadcast::Sender, mpsc};
 
 const MESSAGE_RETRY_API_BASE: &str = "/message_retry";
 
-#[derive(new, Clone)]
+#[derive(new)]
 pub struct MessageRetryApi {
-    tx: Sender<MatchingList>,
+    retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryApiState {
+    pub retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryRequest {
+    pub uuid: String,
+    pub pattern: MatchingList,
+    pub transmitter: mpsc::Sender<MessageRetryResponse>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct MessageRetryResponse {
+    /// ID of the retry request
+    pub uuid: String,
+    /// how many pending operations were evaluated
+    pub evaluated: usize,
+    /// how many of the pending operations matched the retry request pattern
+    pub matched: u64,
 }
 
 async fn retry_message(
-    State(tx): State<Sender<MatchingList>>,
+    State(state): State<MessageRetryApiState>,
     Json(retry_req_payload): Json<MatchingList>,
-) -> String {
-    match tx.send(retry_req_payload) {
-        Ok(_) => "Moved message(s) to the front of the queue".to_string(),
-        // Technically it's bad practice to print the error message to the user, but
-        // this endpoint is for debugging purposes only.
-        Err(err) => format!("Failed to send retry request to the queue: {}", err),
+) -> Result<Json<MessageRetryResponse>, String> {
+    let uuid = uuid::Uuid::new_v4();
+    let uuid_string = uuid.to_string();
+
+    tracing::debug!(uuid = uuid_string, "Sending message retry request");
+
+    // This channel is only created to service this single
+    // retry request so we're expecting a single response
+    // from each transmitter end, hence we are using a channel of size 1
+    let (transmitter, mut receiver) = mpsc::channel(1);
+    state
+        .retry_request_transmitter
+        .send(MessageRetryRequest {
+            uuid: uuid_string.clone(),
+            pattern: retry_req_payload,
+            transmitter,
+        })
+        .map_err(|err| {
+            // Technically it's bad practice to print the error message to the user, but
+            // this endpoint is for debugging purposes only.
+            format!("Failed to send retry request to the queue: {}", err)
+        })?;
+
+    let mut resp = MessageRetryResponse {
+        uuid: uuid_string,
+        evaluated: 0,
+        matched: 0,
+    };
+
+    // Wait for responses from relayer
+    tracing::debug!(uuid = resp.uuid, "Waiting for response from relayer");
+    while let Some(relayer_resp) = receiver.recv().await {
```

Comment:
this doesn't guarantee that it waits for all responses from all the submitters right?
e.g. if like 2 out of 5 responses come in and we receive those 2 but still have 3 responses that haven't come in yet, we never get those 3 responses

Reply:
Wonder if we could use the transmitter's https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Sender.html#method.receiver_count to figure out how many responses we'd expect to get back? and only move on after we've gotten `receiver_count` responses

Reply:
chatted with @kamiyaa about this, elsewhere in the codebase we use `try_recv` which behaves as you describe, but [recv](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.recv) only returns `None` when the channel is closed according to the docs (i.e. when all transmitters are dropped)

Reply:
funnily enough we chatted about that exact approach of counting submitters and responses but just using `recv` is simpler

Reply:
> e.g. if like 2 out of 5 responses come in and we receive those 2 but still have 3 responses that haven't come in yet, we never get those 3 responses

I think in this case, the function will block until those 3 responses come in? Is that not the behavior we want?

Reply:
ahh nice I didn't read the docs, good call this is way better